coderd/prebuilds/reconcile.go

```diff
@@ -0,0 +1,110 @@
+type ReconciliationActions struct {
+	// ActionType determines which field is set and what action should be taken
+	ActionType ActionType
+
+	// Create is set when ActionType is ActionTypeCreate and indicates the number of prebuilds to create
+	Create int32
+
+	// DeleteIDs is set when ActionType is ActionTypeDelete and contains the IDs of prebuilds to delete
+	DeleteIDs []uuid.UUID
+
+	// BackoffUntil is set when ActionType is ActionTypeBackoff and indicates when to retry creating prebuilds
+	BackoffUntil time.Time
+}
+
+type ReconciliationState struct {
+	Actual     int32 // Number of currently running prebuilds
+	Desired    int32 // Number of prebuilds desired as defined in the preset
+	Eligible   int32 // Number of prebuilds that are ready to be claimed
+	Extraneous int32 // Number of extra running prebuilds beyond the desired count
+
+	// Counts of prebuilds in various transition states
+	Starting int32
+	Stopping int32
+	Deleting int32
+}
+
+func NewGlobalSnapshot(presets []database.GetTemplatePresetsWithPrebuildsRow, runningPrebuilds []database.GetRunningPrebuiltWorkspacesRow,
+	prebuildsInProgress []database.CountInProgressPrebuildsRow, backoffs []database.GetPresetsBackoffRow,
+) GlobalSnapshot {
+	return GlobalSnapshot{Presets: presets, RunningPrebuilds: runningPrebuilds, PrebuildsInProgress: prebuildsInProgress, Backoffs: backoffs}
+}
+
+func (s GlobalSnapshot) FilterByPreset(presetID uuid.UUID) (*PresetSnapshot, error) {
+	preset, found := slice.Find(s.Presets, func(preset database.GetTemplatePresetsWithPrebuildsRow) bool {
+		return preset.ID == presetID
+	})
+	if !found {
+		return nil, xerrors.Errorf("no preset found with ID %q", presetID)
+	}
+
+	running := slice.Filter(s.RunningPrebuilds, func(prebuild database.GetRunningPrebuiltWorkspacesRow) bool {
+		if !prebuild.CurrentPresetID.Valid {
+			return false
+		}
+		return prebuild.CurrentPresetID.UUID == preset.ID &&
+			prebuild.TemplateVersionID == preset.TemplateVersionID // Not strictly necessary since presets are 1:1 with template versions, but no harm in being extra safe.
+	})
+
+	// These aren't preset-specific, but they need to inhibit all presets of this template from operating since they could
+	// be in-progress builds which might impact another preset. For example, if a template goes from no defined prebuilds to defined prebuilds
+	// and back, or a template is updated from one version to another.
+	// We group by the template so that all prebuilds being provisioned for a prebuild are inhibited if any prebuild for
+	// any preset in that template are in progress, to prevent clobbering.
+	inProgress := slice.Filter(s.PrebuildsInProgress, func(prebuild database.CountInProgressPrebuildsRow) bool {
+		return prebuild.TemplateID == preset.TemplateID
+	})
+
+	var backoff *database.GetPresetsBackoffRow
+	backoffs := slice.Filter(s.Backoffs, func(row database.GetPresetsBackoffRow) bool {
+		return row.PresetID == preset.ID
+	})
+	if len(backoffs) == 1 {
```

Comment:
is it guaranteed that len(backoffs) is always zero or one?

Reply:
I fixed it like this:
```go
var backoffPtr *database.GetPresetsBackoffRow
backoff, found := slice.Find(s.Backoffs, func(row database.GetPresetsBackoffRow) bool {
	return row.PresetID == preset.ID
})
if found {
	backoffPtr = &backoff
}
```

Now I think we can do the same for `InProgress`. We always have `0` or `1` `InProgress` record per preset.